commerce-cart-content-web/src/main/java/com/liferay/commerce/cart/content/web/internal/portlet/action/AddCommerceOrderItemMVCActionCommand.java

```diff
@@ -113,10 +113,28 @@ protected void doProcessAction(
 			ServiceContext serviceContext = ServiceContextFactory.getInstance(
 				CommerceOrderItem.class.getName(), httpServletRequest);
 
-			CommerceOrderItem commerceOrderItem =
-				_commerceOrderItemService.addCommerceOrderItem(
-					commerceOrder.getCommerceOrderId(), cpInstanceId, quantity,
-					0, ddmFormValues, commerceContext, serviceContext);
+			CommerceOrderItem commerceOrderItem = null;
+
+			List<CommerceOrderItem> cpInstanceCommerceOrderItems =
```

Comment:
@ethanbustad I think we should change this. Doing like this we are making the check in a different db transaction compared to the add or update. 

Maybe is a better idea to move this logic in the addCommerceOrderItem in the local service?  

What do you think?

Reply:
Mmmm good point.... this definitely has concurrency problems.

I don't remember why, but I didn't like the idea of putting the logic in `addCommerceOrderItem` -- maybe just because it's not necessarily _adding_ a new item.

But you're right, this needs to change. I'll take another look.